### PR TITLE
update contributing guide

### PR DESCRIPTION
**Pull request by clewis7**
Start to update the contributing guide

- [x] more detailed instructions on contributing
- [x] move specific information about fpl API to developer notes section in the docs
     - [x] graphics
     - [x] graphic features
     - [x] layouts 

will wait to add developer notes for other parts of the API later when they are more stable

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,221 +1,295 @@
...
-cd fastplotlib
+git remote add upstream https://github.com/fastplotlib/fastplotlib
+```
+
+At this point you have two remotes: `origin` (your fork) and `upstream` (the canonical version). You won't have permission to push to upstream (only `origin`), but 
+this make it easy to keep your `fastplotlib` up-to-date with the canonical version by pulling from upstream: `git pull upstream`.
+
+### Creating a new branch
 
+As mentioned previously, each feature in `fastplotlib` is worked on in a separate branch. This allows multiple people developing multiple features simultaneously, without interfering with each other's work. To create
```
Comment (kushalkolar) [inline CONTRIBUTING.md:79]:
```suggestion
As mentioned previously, each feature in `fastplotlib` is worked on in a separate branch. This allows multiple people to develop multiple features simultaneously, without interfering with each other's work. To create
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,221 +1,295 @@
...
+- A summary including information on what you changed and why
+- References to relevant issues or discussions
+- Special notice to any portion of your changes where you have lingering questions (e.g., "was this the right way to implement this?") or
+want reviewers to pay special attention to
 
-A `Graphic` is something that can be added to a `PlotArea` (described in detail in a later section). All the various 
-fastplotlib graphics, such as `ImageGraphic`, `ScatterGraphic`, etc. inherit from the `Graphic` base class in 
-`fastplotlib/graphics/_base.py`. It has a few properties that mostly wrap `pygfx` `WorldObject` properties and transforms. 
-These might change in the future (ex. `Graphic.position_x` etc.).
+Next, we will be notified of the pull request and will read it over. We will try to give and initial response quickly, and then do a longer in-depth
```
Comment (kushalkolar) [inline CONTRIBUTING.md:144]:
```suggestion
Next, we will be notified of the pull request and will read it over. We will try to give an initial response quickly, and then do a longer in-depth
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,221 +1,295 @@
...
+review, at which point you will probably need to respond to our comments, making changes as appropriate. We will then respond again, and proceed
+in an iterative fashion until everyone is happy with the proposed changes.
 
-All graphics can be given a string name for the user's convenience. This allows graphics to be easily accessed from 
-plots, ex: `subplot["some_image"]`.
+Once your changes are integrated, you will be added as a GitHub contributor. Thank you for being 
+apart of `fastplotlib`!
```
Comment (kushalkolar) [inline CONTRIBUTING.md:149]: "a part"

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,221 +1,295 @@
...
-#### Graphic properties
+As far as code style, please adhere to the following guidelines:
 
-Graphic properties are all evented, and internally we called these "graphic features". They are the various 
-aspects of a graphic that the user can change.
-The "graphic features" subpackage can be found at `fastplotlib/graphics/_features`. As we can see this 
-is a private subpackage and never meant to be accessible to users..
+- Longer, descriptive names are preferred (e.g., `x` is not an appropriate name for a variable), especially for anything user-facing,
+such as methods, attributes, or arguments
+- Any public method or function must have complete type-annotated docstrings (see below for details). Hidden ones do not need to have
```
Comment (kushalkolar) [inline CONTRIBUTING.md:157]:
```suggestion
- Any public method, property, or attribute must have complete type-annotated docstrings (see below for details). Private methods or attributes do not need to have a
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,221 +1,295 @@
...
+complete docstring, but they probably should.
 
-##### LineGraphic
+### Releases
 
-For example let's look at `LineGraphic` in `fastplotlib/graphics/line.py`. Every graphic has a class variable called 
-`_features` which is a set of all graphic properties that are evented. It has the following evented properties: 
-`"data", "colors", "cmap", "thickness"` in addition to properties common to all graphics, such as `"name", "offset", "rotation", and "visible"` 
+We create releases on GitHub, deploy on / distribute via [pypi](https://pypi.org/), and try to follow [semantic versioning](https://semver.org/):
```
Comment (kushalkolar) [inline CONTRIBUTING.md:162]:
```suggestion
We create releases on GitHub and distribute via [pypi](https://pypi.org/), and try to follow [semantic versioning](https://semver.org/):
```

Comment (kushalkolar) [reply CONTRIBUTING.md:162]: We're currently in alpha so we're not exactly following this, we're still at zero version

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -1,221 +1,295 @@
...
+> 2. MINOR version when you add functionality in a backward compatible manner
+> 3. PATCH version when you make backward compatible bug fixes
 
-`VertexColors`behaves similarly, but it can perform additional parsing that can create the colors buffer from different 
-forms of user input. For example if a user runs: `line_graphic.colors = "blue"`, then `VertexColors.__setitem__()` will 
-create a buffer that corresponds to what `pygfx.Color` thinks is "blue". Users can also take advantage of fancy indexing, 
-ex: `line_graphics.colors[bool_array] = "red"` :smile: 
+To release a new version, we [create a GitHub release](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository) with a new tag incrementing the version as described above. 
+Creating the GitHub release will trigger the deployment to pypi, via our `deploy` action (found in `.github/workflows/pypi-publish.yml`). 
+The built version will grab the version tag from the GitHub release, using [setuptools_scm](https://github.com/pypa/setuptools_scm).
```
Comment (kushalkolar) [inline CONTRIBUTING.md:171]: bro even I don't know this XD

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,221 +1,295 @@
...
+### Testing 
 
-`LineGraphic` also has a `thickness` property which is pretty simple, and `DeletedFeature` which is useful if you need 
-callbacks to indicate that the graphic has been deleted (for example, removing references to a graphic from a legend).
+#### Testing Details
 
-Other graphics have properties that are relevant to them, for example `ImageGraphic` has `cmap`, `vmin`, `vmax`,
-properties unique to images.
+An integral part of our testing suite is not only to test that our code is working, but to also check that are plotting 
+library CI pipeline is producing things that "look visually correct".
```
Comment (kushalkolar) [inline CONTRIBUTING.md:178]: simplify, something like: "As a plotting library we require two layers of testing. We use a backend test suite that verifies the basic functionality of buffer managers, graphics, layouts, etc., and another test suite which verifies that the library renders plots that are _visually correct_"

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,221 +1,295 @@
...
-`PlotArea` has the following key properties that allow it to be a "plot area" that can be used to view graphical objects:
+1. Download the regenerated screenshots from the [`fastplotlib` GitHub Actions page](https://github.com/fastplotlib/fastplotlib/actions/workflows/screenshots.yml) for your specific PR
 
-* scene - instance of `pygfx.Scene`
-* canvas - instance of `WgpuCanvas`
-* renderer - instance of `pygfx.WgpuRenderer`
-* viewport - instance of `pygfx.Viewport`
-* camera - instance of `pygfx.PerspectiveCamera`, we always just use `PerspectiveCamera` and just set `camera.fov = 0` for orthographic projections
-* controller - instance of `pygfx.Controller`
+2. Replace the screenshots in your local `fastplotlib` screenshots directories with those downloaded
```
Comment (kushalkolar) [inline CONTRIBUTING.md:199]: replace only the screenshots that the PR changes

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -1,221 +1,295 @@
...
+```
 
-* graphics - a tuple of weakref proxies to all `Graphics` within this `PlotArea`, users are only given weakref proxies to `Graphic` objects, all `Graphic` objects are stored in a private global dict.
-* selectors - a tuple of weakref proxies to all selectors within this `PlotArea`
-* legend - a tuple of weakref proxies to all legend graphics within this `PlotArea`
-* name - plot areas are allowed to have names that the user can use for their convenience
+#### Adding tests
 
-Important methods:
+Depending on the type of contribution you are making, new tests will need to be added to the repository. Unit tests for testing underlying functionality such as buffer managers, figure instantiation, and
```
Comment (kushalkolar) [inline CONTRIBUTING.md:222]:
```suggestion
Depending on the type of contribution you are making, new tests might need to be added to the repository. Unit tests for testing underlying functionality such as buffer managers, figure instantiation, and
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -1,221 +1,295 @@
...
 
-* add_graphic - add a `Graphic` to the `PlotArea`, append to the end of the `PlotArea._graphics` list
-* insert_graphic - insert a `Graphic` to the `PlotArea`, insert to a specific position of the `PlotArea._graphics` list
-* remove_graphic - remove a graphic from the `Scene`, **does not delete it**
-* delete_graphic - delete a graphic from the `PlotArea`, performs garbage collection
-* clear - deletes all graphics from the `PlotArea`
-* center_graphic - center camera w.r.t. a `Graphic`
-* center_scene - center camera w.r.t. entire `Scene`
-* auto_scale - Auto-scale the camera w.r.t to the `Scene`
+If you are adding a new example to the library, you will need to add to comments to the top of your `.py` file in order to make sure it is both tested and added to the gallery.
```
Comment (kushalkolar) [inline CONTRIBUTING.md:225]:
```suggestion
If you are adding a new example to the library, you will need the following comments to the top of your `.py` file in order to make sure it is both tested and added to the gallery. 
```

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,221 +1,295 @@
...
 
-`GraphicMethodsMixin` is a simple class that just has all the `add_<graphic>` methods. It is autogenerated by a utility script like this:
+If you would like to build the documentation locally:
 
 ```bash
-python scripts/generate_add_methods.py
-```
+cd docs
+# regenerate the api guide
+python source/generate_api.py
```
Comment (kushalkolar) [inline CONTRIBUTING.md:245]: oh somewhere state that any changes to the API require this to be run before making a PR

Comment (kushalkolar) [reply CONTRIBUTING.md:245]: Also any changes to Graphics constructors will require regenerating the mixin class, that info can go somewhere else :thinking:

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -1,221 +1,295 @@
...
+make html
+```
 
-Subplot has one property that is not in `PlotArea`:
+#### Adding documentation
 
-* docks: a `dict` of `PlotAreas` which are located at the "top", "right", "left", and "bottom" edges of a `Subplot`. By default their size is `0`. They are useful for putting things like histogram LUT tools.
+All public-facing functions and classes should have complete docstrings, which start with a one-line short summary of the function, 
+a medium-length description of the function / class and what it does, and a complete description of all arguments and return values. 
+Docstrings should be relatively short, providing the information necessary for a user to use the code.
```
Comment (kushalkolar) [inline CONTRIBUTING.md:255]:
```suggestion
Docstrings should comprehensive, providing the information necessary for a user to use the method or property without going through the code.
```

Our docstrings are  not short XD, good libs generally have comprehensive docstrings, see sklearn.

**Review by kushalkolar — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 6 above)
Comment (clewis7) [reply CONTRIBUTING.md:171]: TYL

File: docs/source/developer_notes/graphics.rst [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+            └─ LineStack
+
+..
+
+All graphics can be given a string name for the user's convenience. This allows graphics to be easily accessed from
+plots, ex: ``subplot["some_image"]``.
+
+All graphics contain a ``world_object`` property which is just the ``pygfx.WorldObject`` that this graphic uses. Fastplotlib
+keeps a *private* global dictionary of all ``WorldObject`` instances and users are only given a weakref proxy to this world object.
+This is due to garbage collection. This may be quite complicated for beginners, for more details see this PR: https://github.com/fastplotlib/fastplotlib/pull/160 .
```
Comment (kushalkolar) [inline docs/source/developer_notes/graphics.rst:36]: You could add something like "furthermore, garbage collection is even more complicated in ipython and jupyter. If you're curious see: https://github.com/fastplotlib/fastplotlib/pull/546 "

File: docs/source/developer_notes/graphics.rst [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+If you are curious or have more questions on garbage collection in ``fastplotlib`` you're welcome to post an issue :D.
+
+Graphic collections are groups of graphics. For now, we have a ``LineCollection`` which is a group of ``LineGraphic`` objects. We also have a ``LineStack`` which
```
Comment (kushalkolar) [inline docs/source/developer_notes/graphics.rst:39]:
```suggestion
Graphic collections are collections of graphics. For now, we have a ``LineCollection`` which is a collection of ``LineGraphic`` objects. We also have a ``LineStack`` which
```

So that users don't get confused with `pygfx.Group` which is completely diff.

File: docs/source/developer_notes/graphics.rst [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+inherits from ``LineCollection`` and gives some fixed offset between ``LineGraphic`` objects in the collection.
```
Comment (kushalkolar) [inline docs/source/developer_notes/graphics.rst:40]: Add something like "A graphic collection behaves kind of like an array of graphics"

File: docs/source/developer_notes/graphics.rst [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+aspects of a graphic that the user can change.
+The "graphic features" subpackage can be found at ``fastplotlib/graphics/_features``. As we can see this
+is a private subpackage and never meant to be accessible to users.
+
+For example let's look at ``LineGraphic`` in ``fastplotlib/graphics/line.py``. Every graphic has a class variable called
+``_features`` which is a set of all graphic properties that are evented. It has the following evented properties:
+``"data", "colors", "cmap", "thickness"`` in addition to properties common to all graphics, such as ``"name", "offset", "rotation", and "visible"``
+
+Now look at the constructor for the ``LineGraphic`` base class ``PositionsGraphic``, it first creates an instance of ``VertexPositions``.
+This is a class that manages vertex positions buffer. It defines the line, and provides additional useful functionality.
```
Comment (kushalkolar) [inline docs/source/developer_notes/graphics.rst:55]:
```
This is a class that manages vertex positions buffer. For the user, it defines the line data, and provides additional useful functionality.
```

File: docs/source/developer_notes/layouts.rst [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+
+``PlotArea`` has the following key properties that allow it to be a "plot area" that can be used to view graphical objects:
+
+* scene - instance of ``pygfx.Scene``
+* canvas - instance of ``WgpuCanvas``
+* renderer - instance of ``pygfx.WgpuRenderer``
+* viewport - instance of ``pygfx.Viewport``
+* camera - instance of ``pygfx.PerspectiveCamera``, we always just use ``PerspectiveCamera`` and just set ``camera.fov = 0`` for orthographic projections
+* controller - instance of ``pygfx.Controller``
+
```
Comment (kushalkolar) [inline docs/source/developer_notes/layouts.rst:18]: add something like: "If these concepts are unfamiliar to you we recommend learning about how rendering engines work, the pygfx guide is a great place to start! https://docs.pygfx.org/stable/guide.html"

Comment (kushalkolar) [reply docs/source/developer_notes/layouts.rst:18]: This is also a short quick video that goes through the basic concepts: https://www.youtube.com/watch?v=cvcAjgMUPUA

File: docs/source/developer_notes/layouts.rst [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+This class inherits from ``PlotArea`` and ``GraphicMethodsMixin``.
+
+``GraphicMethodsMixin`` is a simple class that just has all the ``add_<graphic>`` methods. It is autogenerated by a utility script like this:
+
+.. code-block:: bash
+
+    python scripts/generate_add_methods.py
+
+Each ``add_<graphic>`` method basically creates an instance of ``Graphic``, adds it to the ``Subplot``, and returns a weakref
```
Comment (kushalkolar) [inline docs/source/developer_notes/layouts.rst:62]:
```suggestion
Each ``add_<graphic>`` method basically creates an instance of ``Graphic``, adds it to the ``Subplot``, and returns the ``Graphic``
```

graphics are no longer weakrefs, only worldobjects still are

File: docs/source/developer_notes/layouts.rst [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+proxy to the ``Graphic``.
```
Comment (kushalkolar) [inline docs/source/developer_notes/layouts.rst:63]:
```suggestion
```

File: docs/source/developer_notes/layouts.rst [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+Subplot has one property that is not in ``PlotArea``:
```
Comment (kushalkolar) [inline docs/source/developer_notes/layouts.rst:65]:
```suggestion
``Subplot`` has one property that is not in ``PlotArea``:
```

**Review by kushalkolar — CHANGES_REQUESTED** [review]
few very minor things and then good to go! :partying_face:

**Comment by kushalkolar** [comment]
@clewis7 after addressing these and once #639 is in main we can rebase this onto main. And then what do you think about add a few base classes like `Graphic`, `FeatureEvent`, `BufferManager` etc to the API docs since that's useful for devs.

**Comment by clewis7** [comment]
superseded by #667
